tools/pythonpkg/src/pyconnection.cpp

```diff
@@ -597,25 +627,54 @@ shared_ptr<DuckDBPyConnection> DuckDBPyConnection::Connect(const string &databas
 		}
 		config.SetOption(*config_property, Value(val));
 	}
-	res->database = make_unique<DuckDB>(database, &config);
-	res->connection = make_unique<Connection>(*res->database);
+
 	if (config.enable_external_access) {
-		res->database->instance->config.replacement_scans.emplace_back(ScanReplacement);
-	}
+		config.replacement_scans.emplace_back(ScanReplacement);
+	}
+	if (abs_path == ":memory:") {
+		new_instance = true;
+		return make_shared<DuckDB>(database, &config);
+	}
+	shared_ptr<DuckDB> db_instance;
+	if (db_instances.find(abs_path) != db_instances.end()) {
+		db_instance = db_instances[abs_path].lock();
+		if (db_instance) {
+			auto existing_context = db_instance->instance->GetConnectionManager().GetConnectionList().front().get();
+			if (!IsConfigurationSame(existing_context, db_instance->instance->config, config_dict, read_only)) {
+				throw std::runtime_error("Can't open a connection to same database file with a different configuration "
```

Comment:
Maybe add the database path to the error message, and clarify that another open connection exists?